### PR TITLE
Xcode 9: Automatically detect selected provisioning profiles of the Xcode project

### PR DESCRIPTION
**Pull request by KrauseFx**
This way fastlane will just keep working with Xcode 9 even for setups that don't use _match_ (yet).

Xcode 9 requires a mapping of bundle identifier and what provisioning profile to use. We use `xcodeproj` to find those values, but in a way that _gym_ doesn't break if we can't detect those things

I also verified that things work fine if you pass non-existent bundle identifier/profiles, Xcode will just ignore them 👍

Fixes https://github.com/fastlane/fastlane/issues/9380 even more

Output when running it:
screen shot 2017-06-13 at 2 46 41 pm

Error message if it's not auto-detected, catching Xcode error, and linking to the new docs :) 
screen shot 2017-06-13 at 2 41 55 pm

### CODE REVIEW

File: fastlane_core/lib/fastlane_core/project.rb [hunk 1]
```diff
@@ -292,7 +292,7 @@ def build_xcodebuild_showbuildsettings_command
     end
 
     # Get the build settings for our project
-    # this is used to properly get the DerivedData folder
+    # this is used e.g. to properly get the DerivedData folder
```
Comment (ohayon) [inline fastlane_core/lib/fastlane_core/project.rb]: I think this line should be reworded to just say `e.g. to properly get the DerivedData folder`

File: gym/README.md [hunk 2]
```diff
@@ -174,6 +174,20 @@ export_options(
 )
 ```
 
+Optional: To pass a mapping of bundle identifiers to provisioning profiles, use the following
```
Comment (ohayon) [inline gym/README.md]: Should we add another comment here about why someone would want to use this?

File: gym/lib/gym/error_handler.rb [hunk 3]
```diff
@@ -80,6 +80,10 @@ def handle_package_error(output)
           print "This means, the specified provisioning profile was not created using"
           print "the specified certificate."
           print "Run cert and sigh before gym to make sure to have all signing resources ready"
+        when /requires a provisioning profile/
+          print "No provisioning profile provided"
+          print "Make sure to pass a valid provisioning for each rquired target"
```
Comment (ohayon) [inline gym/lib/gym/error_handler.rb]: *required

File: gym/lib/gym/detect_values.rb [hunk 4]
```diff
@@ -62,6 +63,42 @@ def self.xcode_preferences_dictionary(path)
...
+          bundle_identifier = current["PRODUCT_BUNDLE_IDENTIFIER"]
+          provisioning_profile_specifier = current["PROVISIONING_PROFILE_SPECIFIER"]
+          next if provisioning_profile_specifier.to_s.length == 0
+
+          provisioning_profile_mapping[bundle_identifier] = provisioning_profile_specifier
+        end
+      end
+
+      Gym.config[:export_options] ||= {}
+      Gym.config[:export_options][:provisioningProfiles] = provisioning_profile_mapping
```
Comment (ohayon) [inline gym/lib/gym/detect_values.rb]: Does this overwrite any export options that are already set?

**Review by ohayon — CHANGES_REQUESTED** [review]
This looks great! Just a couple of questions, but otherwise 💯

**Comment by fastlane-bot-helper** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Tests have failed, see below for more information.</td>
    </tr>
  </tbody>
</table>

### Tests: 

Name | File|
--- | ---|
Gym Gym::PackageCommandGeneratorXcode7 generates a valid plist file we need | <a href='https://github.com/fastlane/fastlane/blob/cf763ea9dbec526bcc9d37d90cdff7d59cb7db48/./gym/spec/package_command_generator_xcode7_spec.rb'>package_command_generator_xcode7_spec.rb</a>|
Gym Gym::PackageCommandGeneratorXcode7 reads export options from hash | <a href='https://github.com/fastlane/fastlane/blob/cf763ea9dbec526bcc9d37d90cdff7d59cb7db48/./gym/spec/package_command_generator_xcode7_spec.rb'>package_command_generator_xcode7_spec.rb</a>|
Gym Gym::PackageCommandGeneratorXcode7 doesn't store bitcode/symbols information for non app-store builds | <a href='https://github.com/fastlane/fastlane/blob/cf763ea9dbec526bcc9d37d90cdff7d59cb7db48/./gym/spec/package_command_generator_xcode7_spec.rb'>package_command_generator_xcode7_spec.rb</a>|


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Review by ohayon — APPROVED** [review]

**Comment by ohayon** [comment]
Does this PR take care of this change too? https://github.com/fastlane/fastlane/pull/9449

**Comment by KrauseFx** [comment]
Nice catch @ohayon, seems like it's the same thing, but with a different approach. Let's make sure we use the better one 👍

**Comment by igor-makarov** [comment]
As I mentioned in [#9449](https://github.com/fastlane/fastlane/pull/9449), I like combining both approaches as they do not seem contradictive of each other.

I do want to report a small bug I encountered while testing this PR combined with mine:

When using `workspace` instead of `project` in gym options, gym fails to find the project file to infer from. The error I get is:
```
The plist file at path `/Users/*****/*****/*****.xcworkspace/project.pbxproj` doesn't exist.
```

**Comment by andrerfneves** [comment]
@KrauseFx when do you believe I can expect this to be merged in?

**Comment by KrauseFx** [comment]
Whops, I thought this was already merged, thanks for pinging me. Looking into this now 👍

**Comment by KrauseFx** [comment]
Thanks for letting me know re workspace, I can reproduce @igor-makarov 👍

**Comment by KrauseFx** [comment]
@igor-makarov Would you mind giving it another shot for your workspace?

**Comment by igor-makarov** [comment]
Still not working on my config... Two issues: 

- Iterating on the workspace can produce groups other than a project (in my case, a GPX folder). My suggestion is to add `current_match.end_with?(".xcodeproj")` to the `find_all` block.
- Not related to workspaces, but I have several base `.xcconfig` files that the project build settings depend on. The provisioning profile is specified there, like so:
```
PROVISIONING_PROFILE_SPECIFIER = ${PRODUCT_BUNDLE_IDENTIFIER} AppStore
```
Because of that, in my setup, `current["PRODUCT_BUNDLE_IDENTIFIER"]` returns nil.

However, attempting to use [Xcodeproj](https://github.com/CocoaPods/Xcodeproj) method `target.resolved_build_setting` function doesn't seem to work as it returns the value resolved from the xcconfig, but doesn't resolve the fact the value references another property, in this case `"$(XXXXXX_BASE_BUNDLE_PREFIX).Widget"`.

This seems to be a limitation of Xcodeproj and I don't know how easy it is to address it. Perhaps @dantoml can clarify what can be done about this?

**Comment by andrerfneves** [comment]
@KrauseFx any updates on getting this merged?

**Comment by KrauseFx** [comment]
Yeah, it would be awesome if you could provide a sample project that can reproduce the problem.

@andrerfneves we'll merge as soon as it's ready 😉 it seems like we still have some problems with workspaces.

I'm not gonna be online tomorrow (Monday), so realistically I'll have time to finish this PR on Tuesday 👍

**Comment by KrauseFx** [comment]
For now, you can always manually pass the provisioning profile mapping as described in the docs @andrerfneves

**Comment by andrerfneves** [comment]
@KrauseFx Ah! wasn't aware I could do that. Sorry about that. And thanks for the hard work on such an awesome tool! Cheers

**Comment by KrauseFx** [comment]
Without a sample project to reproduce the issue, I'm having a hard time fixing the problem 😢

**Comment by igor-makarov** [comment]
I can provide one. Any reason not to merge the pull request I submitted, by the way? It doesn't conflict with this one

**Comment by igor-makarov** [comment]
I've created the demo project. You can find it at https://github.com/igor-makarov/DemoWorkspaceWithXCConfig

Note that the issue I'm running into isn't related to workspaces. It's related to inherited `.xccconfig` files.

Also, my other issue (Iterating on the workspace can produce groups other than a project) that I've mentioned above, still needs to be addressed.

**Comment by KrauseFx** [comment]
@igor-makarov Thanks for preparing the sample project, I gave it a try by changing the bundle identifier to my project one, and things are working without any problems. 

```
[18:28:58]: Detected provisioning profile mapping: {"family.wwdc.app"=>"match AppStore family.wwdc.app"}
[18:28:58]: $ xcodebuild -list -workspace ./DemoWorkspaceWithXCConfig.xcworkspace
[18:28:59]: $ xcodebuild -showBuildSettings -workspace ./DemoWorkspaceWithXCConfig.xcworkspace -scheme DemoWorkspaceWithXCConfig

+-----------------------------------------------------+---------------------------------------------------------+
|                                            Summary for gym 2.38.1                                             |
+-----------------------------------------------------+---------------------------------------------------------+
| workspace                                           | ./DemoWorkspaceWithXCConfig.xcworkspace                 |
| export_options.provisioningProfiles.family.wwdc.app | match AppStore family.wwdc.app                          |
| scheme                                              | DemoWorkspaceWithXCConfig                               |
| destination                                         | generic/platform=iOS                                    |
| output_name                                         | DemoWorkspaceWithXCConfig                               |
| build_path                                          | /Users/fkrause/Library/Developer/Xcode/Archives/2017-0  |
|                                                     | 6-29                                                    |
| clean                                               | false                                                   |
| output_directory                                    | .                                                       |
| silent                                              | false                                                   |
| skip_package_ipa                                    | false                                                   |
| use_legacy_build_api                                | false                                                   |
| buildlog_path                                       | ~/Library/Logs/gym                                      |
| xcode_path                                          | /Applications/Xcode-9.app                               |
+-----------------------------------------------------+---------------------------------------------------------+

[18:29:00]: $ set -o pipefail && xcodebuild -workspace ./DemoWorkspaceWithXCConfig.xcworkspace -scheme DemoWorkspaceWithXCConfig -destination 'generic/platform=iOS' -archivePath /Users/fkrause/Library/Developer/Xcode/Archives/2017-06-29/DemoWorkspaceWithXCConfig\ 2017-06-29\ 18.29.00.xcarchive archive | tee /Users/fkrause/Library/Logs/gym/DemoWorkspaceWithXCConfig-DemoWorkspaceWithXCConfig.log | xcpretty
[18:29:01]: ▸ Building DemoWorkspaceWithXCConfig/DemoWorkspaceWithXCConfig [Release]
[18:29:01]: ▸ Check Dependencies
[18:29:01]: ▸ Linking DemoWorkspaceWithXCConfig
[18:29:02]: ▸ Compiling LaunchScreen.storyboard
[18:29:05]: ▸ Compiling Main.storyboard
[18:29:05]: ▸ Processing Info.plist
[18:29:05]: ▸ Generating 'DemoWorkspaceWithXCConfig.app.dSYM'
[18:29:05]: ▸ Touching DemoWorkspaceWithXCConfig.app
[18:29:10]: ▸ Signing /Users/fkrause/Library/Developer/Xcode/DerivedData/DemoWorkspaceWithXCConfig-cfpacxhsmflleecusxxxcgiuiega/Build/Intermediates.noindex/ArchiveIntermediates/DemoWorkspaceWithXCConfig/InstallationBuildProductsLocation/Applications/DemoWorkspaceWithXCConfig.app
[18:29:11]: ▸ Touching DemoWorkspaceWithXCConfig.app.dSYM
[18:29:11]: ▸ Archive Succeeded
[18:29:11]: Generated plist file with the following values:
[18:29:11]: ▸ -----------------------------------------
[18:29:11]: ▸ {
[18:29:11]: ▸   "provisioningProfiles": {
[18:29:11]: ▸     "family.wwdc.app": "match AppStore family.wwdc.app"
[18:29:11]: ▸   },
[18:29:11]: ▸   "method": "app-store"
[18:29:11]: ▸ }
[18:29:11]: ▸ -----------------------------------------
[18:29:11]: $ /usr/bin/xcrun /Users/fkrause/Developer/fastlane/gym/lib/assets/wrap_xcodebuild/xcbuild-safe.sh -exportArchive -exportOptionsPlist '/var/folders/8q/f0rt7xtj4pd8q6vgh_mk2p4c00dw80/T/gym_config20170629-19615-3ens69.plist' -archivePath /Users/fkrause/Library/Developer/Xcode/Archives/2017-06-29/DemoWorkspaceWithXCConfig\ 2017-06-29\ 18.29.00.xcarchive -exportPath '/var/folders/8q/f0rt7xtj4pd8q6vgh_mk2p4c00dw80/T/gym_output20170629-19615-1mhwq53'
[18:29:35]: Compressing 1 dSYM(s)
[18:29:35]: $ cd '/Users/fkrause/Library/Developer/Xcode/Archives/2017-06-29/DemoWorkspaceWithXCConfig 2017-06-29 18.29.00.xcarchive/dSYMs' && zip -r '/Users/fkrause/Developer/hacking/DemoWorkspaceWithXCConfig/DemoWorkspaceWithXCConfig.app.dSYM.zip' *.dSYM
[18:29:35]: ▸   adding: DemoWorkspaceWithXCConfig.app.dSYM/ (stored 0%)
[18:29:35]: ▸   adding: DemoWorkspaceWithXCConfig.app.dSYM/Contents/ (stored 0%)
[18:29:35]: ▸   adding: DemoWorkspaceWithXCConfig.app.dSYM/Contents/Info.plist (deflated 52%)
[18:29:35]: ▸   adding: DemoWorkspaceWithXCConfig.app.dSYM/Contents/Resources/ (stored 0%)
[18:29:35]: ▸   adding: DemoWorkspaceWithXCConfig.app.dSYM/Contents/Resources/DWARF/ (stored 0%)
[18:29:35]: ▸   adding: DemoWorkspaceWithXCConfig.app.dSYM/Contents/Resources/DWARF/DemoWorkspaceWithXCConfig (deflated 68%)

[18:29:35]: Successfully exported and compressed dSYM file
[18:29:35]: Successfully exported and signed the ipa file:
[18:29:35]: /Users/fkrause/Developer/hacking/DemoWorkspaceWithXCConfig/DemoWorkspaceWithXCConfig.ipa
```

I'll dive in further to figure out what it is.

**Comment by KrauseFx** [comment]
Alright, seems like things work for me, we can do a follow-up patch if things end up not working

**Comment by igor-makarov** [comment]
It seems that it matched the profiles from the `match` output, rather than from the project file.

**Comment by igor-makarov** [comment]
FYI I still think it would fail under my config when not using `match` or the `sigh` patch I submitted. However I'm using sigh so my project builds ok. 

So I guess I'm commenting just in case someone googles xcconfig in the future.

**Comment by mjholgate** [comment]
@KrauseFx I've noticed this still fails when either the provisioning profile or the app bundle ID are inherited from the project configuration (rather than being defined in the target) and are constructed using build setting $(VARIABLES) - this causes values to be missing from the <dict> in the generated plist.

If I've not explained this well enough, please let me know and am happy to produce an example project tomorrow :). Cheers

**Comment by igor-makarov** [comment]
@mjholgate This is my setup exactly and I already linked to a demo before.

**Comment by KrauseFx** [comment]
Yes, I couldn't reproduce, if you let me know how I can reproduce the problem with the sample project, I can work on a fix 👍

**Comment by igor-makarov** [comment]
@KrauseFx I've provided the sample project earlier. 

The reason you haven't encountered the problem when reproducing is because you have `match` in place.

**Pull request merged**